### PR TITLE
Don't spill more blocks than we need to

### PR DESCRIPTION
**Pull request by sryza**
Also fixes a memory leak

### CODE REVIEW

**Comment by jerryshao** [comment]
Hi Sandy, is this code finished? I've seen any code put shuffle blocks into queue. Sorry if I miss anything.

**Comment by sryza** [comment]
Oops, I had meant it to be finished, but looks like I missed that.  Thanks for catching it.  Just uploaded a fix.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala [hunk 1]
```diff
@@ -169,8 +146,53 @@ private[spark] class SortShuffleReader[K, C](
...
+    val fos = new FileOutputStream(file)
+    val bos = new BufferedOutputStream(fos, fileBufferSize)
+
+    // If the remaining unfetched data would fit inside our current allocation, we don't want to
+    // waste time spilling blocks beyond the space needed for it.
+    var bytesToSpill = unfetchedBytes
+    val blocksToSpill = new ArrayBuffer[MemoryShuffleBlock]()
+    blocksToSpill += tippingBlock
+    bytesToSpill -= tippingBlock.blockData.size
+    while (bytesToSpill > 0 && inMemoryBlocks.isEmpty) {
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala]: IIUC, I think it should be `!inMemoryBlocks.isEmpty`.

Comment (sryza) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala]: You are right

**Comment by jerryshao** [comment]
Hi Sandy, thanks for your work. I will merge this.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala [hunk 2]
```diff
@@ -169,8 +146,53 @@ private[spark] class SortShuffleReader[K, C](
...
+    val fos = new FileOutputStream(file)
+    val bos = new BufferedOutputStream(fos, fileBufferSize)
+
+    // If the remaining unfetched data would fit inside our current allocation, we don't want to
+    // waste time spilling blocks beyond the space needed for it.
+    var bytesToSpill = unfetchedBytes
+    val blocksToSpill = new ArrayBuffer[MemoryShuffleBlock]()
+    blocksToSpill += tippingBlock
+    bytesToSpill -= tippingBlock.blockData.size
+    while (bytesToSpill > 0 && !inMemoryBlocks.isEmpty) {
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala:161]: Hi @sryza , I think we should change here to:

``` scala
if (bytesToSpill > 0) {
  while (!inMemoryBlocks.isEmpty) {
    ....
  }
}
```

Seems `bytesToSpill` will be negative and this loop will be jumped out.

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala:161]: Is your purpose to spill only the number of blocks which fit for the left `unfetchBytes` to stay in memory? Seems I misunderstood it.

**Comment by jerryshao** [comment]
Hi Sandy, through test, I think the `unfetchedBytes` computed from MapOutputTracker is not exactly the same as the map output, internally MapOutputTrack represents this length as a compressed 8 bit byte, this will lose the precision. Since Spark itself only use this size to filter out zero block, so use this to do calculation potentially has bugs.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala (hunk 2 above)
Comment (sryza) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleReader.scala:161]: Right.  That way we don't spill more than we need to.

**Comment by sryza** [comment]
Ah, good point, hadn't thought about that.  I think this will still be OK though.  According to the documentation in MapOutputTracker.compressSize, it will be less than 10% off for sizes up to 35GB.  The code doesn't rely on it being exactly correct - the worst thing that can happen is that we're not entirely optimal - i.e. we spill one more block than necessary or we end up doing an extra small spill.

**Comment by jerryshao** [comment]
According to my test, this assertion `assert(unfetchedBytes == 0)` always failed, not matter the map output size, in most situation this compressedSize is slightly different from the original size, because our implementation relies on the exact size, so the program always failed. I think we should figure out another solution to address this problem.

**Comment by sryza** [comment]
Ah, that makes sense. We should remove that assertion, but I think my point above still stands: the implementation in fact does not rely on the exact size.

**Comment by jerryshao** [comment]
I know that your above point still stands, but I think some code logic should be changed, otherwise it will confuse the users, I will try to improve it. Would you mind reviewing it after I'm done.

**Comment by sryza** [comment]
Sound good.  Happy to review.

**Pull request merged**